### PR TITLE
Move some functions to allow them to be inlined with clang

### PR DESCRIPTION
**Pull request by kouvel**
For the moment, GetThread, _rotl/_rotr, Interlocked..., and similar functions were moved to allow inlining them with clang.

The plan is to eventually enable link-time optimization (LTO) to cover inlining across translation units.

Renamed some code files in PAL tests to compile as C++ to use the C++ inline semantics.

### CODE REVIEW

**Comment by kouvel** [comment]
@sergiy-k @jkotas

File: src/vm/threads.cpp [hunk 1]
```diff
@@ -328,11 +328,6 @@ GVAL_IMPL_INIT(DWORD, gAppDomainTLSIndex, TLS_OUT_OF_INDEXES);   // index ( (-1)
 
 #ifndef DACCESS_COMPILE
 #ifdef FEATURE_IMPLICIT_TLS
-EXTERN_C Thread* STDCALL GetThread()
-{
-    return gCurrentThreadInfo.m_pThread;
-}
-
 EXTERN_C AppDomain* STDCALL GetAppDomain()
```
Comment (jkotas) [inline src/vm/threads.cpp]: You should move GetAppDomain too while you are on it.

**Comment by jkotas** [comment]
LGTM

File: src/vm/threads.cpp (hunk 1 above)
Comment (kouvel) [inline src/vm/threads.cpp]: Done

**Comment by sergiy-k** [comment]
LGTM

**Pull request merged**